java/src/main/java/com/genexus/webpanels/GXObjectUploadServices.java

```diff
@@ -84,35 +73,37 @@ protected void doExecute(HttpContext context) throws Exception
 			else
 			{
 				String contentType = context.getHeader("Content-Type");
-				String ext = getExtension(contentType);
-				String tempFileName = com.genexus.PrivateUtilities.getTempFileName(ext);
-				String filePath = tempFileName;		
-				if (Application.getGXServices().get(GXServices.STORAGE_SERVICE) == null)
-				{
-					filePath = Preferences.getDefaultPreferences().getBLOB_PATH() + tempFileName;		
-				}
-				else
-				{
-					filePath = Preferences.getDefaultPreferences().getBLOB_PATH().replace(java.io.File.separator, "/") + 	tempFileName;
-				}				
+				ext = getExtension(contentType);
+				fileName = com.genexus.PrivateUtilities.getTempFileName("tmp");
+				String filePath = fileDirPath + fileName;
+				fileName = fileName.replaceAll(".tmp", "." + ext);
 				FileItem fileItem = new FileItem(filePath, false, "", context.getRequest().getInputStream());
-				filePath = fileItem.getPath();
-				String keyId = UUID.randomUUID().toString().replace("-","");
-				CacheAPI.files().set(keyId, filePath, CommonUtil.UPLOAD_TIMEOUT );
+				savedFileName = fileItem.getPath();
+				JSONObject jObj = new JSONObject();
+				jObj.put("object_id", HttpUtils.getUploadFileId(keyId));
+				if (!isRestCall) {
 				context.getResponse().setContentType("application/json");
 				context.getResponse().setStatus(201);
 				context.getResponse().setHeader("GeneXus-Object-Id", keyId);
-				JSONObject jObj = new JSONObject();
-				jObj.put("object_id", CommonUtil.UPLOADPREFIX + keyId);
 				context.writeText(jObj.toString());
 				context.getResponse().flushBuffer();
+				}
+				else {
+					String jsonResponse = jObj.toString();
+					builder = Response.status(201).entity(jsonResponse);
+					builder.header("GeneXus-Object-Id", keyId);
+				}
 			}
 		}
 		catch (Throwable e)
 		{
 			context.sendResponseStatus(404, e.getMessage());
 		}
 		finally {
+			if (!savedFileName.isEmpty()){
+				HttpUtils.CacheUploadFile(keyId, savedFileName, fileName, ext);
```

Comment:
When context.isMultipartContent() is true, this code stores in cache only the last savedFileName in the iteration through postedFiles. Move these sentences inside the for and inside the else.

Reply:
I think that there's is not a practical case when more that one file are upload in a request,

Reply:
I agree but just for a consistent code (it already has an iteration on postedFiles)  it would be recommended. I already did it.